Exclude added flash types from action_methods

Bodacious opened the pull request:
Exclude additional flash types from action_methods.

Fixes #33593
Copy of stale PR #33594

When new flash types are added to ApplicationController, these
should not appear in the action_methods set, since they're
not actions. This commit corrects this behaviour.

``` ruby
class MyController < ApplicationController
  add_flash_types :hype
end

# Currently:
MyController.action_methods.include?('hype') # => true

# This commit
MyController.action_methods.include?('hype') # => false
```

eugeneius commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,16 @@
+*   Exclude additional flash types from `ActionController::Base.action_methods`
```
```suggestion
*   Exclude additional flash types from `ActionController::Base.action_methods`.
```

eugeneius commented on actionpack/lib/abstract_controller/base.rb:
```diff
@@ -88,14 +88,13 @@ def internal_methods
       # * <tt>Set</tt> - A set of all methods that should be considered actions.
       def action_methods
         @action_methods ||= begin
+
```
```suggestion
```

eugeneius commented on actionpack/lib/abstract_controller/base.rb:
```diff
@@ -88,14 +88,13 @@ def internal_methods
-            public_instance_methods(false))
-
-          methods.map!(&:to_s)
+            public_instance_methods(false)).map(&:to_s)
```
This regresses a performance optimisation from https://github.com/rails/rails/pull/33693.

```suggestion
            public_instance_methods(false))
            
          methods.map!(&:to_s)
```

Bodacious replied:
Well caught, thank you

eugeneius commented on actionpack/lib/action_controller/metal/flash.rb:
```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
             request.flash[type]
           end
           helper_method(type) if respond_to?(:helper_method)
```
```suggestion
          helper_method(type) if respond_to?(:helper_method)

```

eugeneius commented on actionpack/lib/action_controller/metal/flash.rb:
```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
         end
       end
+
+      # Extend AbstractController::Base::action_methods to remove any added flash types
+      # from the set
```
I think this method is simple enough that we don't need internal documentation for it.

```suggestion
```

eugeneius commented on actionpack/lib/action_controller/metal/flash.rb:
```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
+
+      # Extend AbstractController::Base::action_methods to remove any added flash types
+      # from the set
+      def action_methods #:nodoc:
+        super - _flash_types.map(&:to_s).to_set
+      end
```
It's important that `action_methods` is memoised, since it's called during every request.

We can copy the approach from the `UrlFor` module and use the same instance variable name, so that `clear_action_methods!` still works:

https://github.com/rails/rails/blob/a2cb0b23b7233371d8a2e0f843885dae330da2b7/actionpack/lib/abstract_controller/url_for.rb#L24-L30

```suggestion
      def action_methods #:nodoc:
        @action_methods ||= super - _flash_types.map(&:to_s).to_set
      end
```

eugeneius commented:
Can you squash your commits too?

Bodacious committed:
Exclude added flash types from action_methods

Bodacious commented:
@eugeneius I've applied the suggested changes. 

I rebased my PR against main and squashed the commits. Sorry, I wasn't sure if this project squashes on merge or not.

p8 commented:
@Bodacious Could you please rebase on main again?
The spellcheck should be fixed, making this build green again.

eugeneius commented on actionpack/lib/abstract_controller/base.rb:
```diff
@@ -96,7 +97,6 @@ def action_methods
             public_instance_methods(false))
 
           methods.map!(&:to_s)
-
```
There are still unnecessary changes in this file - we try to avoid those to keep the git history clean.

Can you also clean up the commit message, it contains all of the squashed commits' messages (e.g. "Add suggested changes from feedback").

Bodacious commented:
@eugeneius Thanks. I've removed that extra whitespace and cleaned up the commit message

eugeneius commented on actionpack/lib/abstract_controller/base.rb:
```diff
@@ -96,7 +96,6 @@ def action_methods
             public_instance_methods(false))
 
           methods.map!(&:to_s)
-
```
Can you restore this line break? Otherwise I think this is good to go.

Bodacious replied:
Sorry, I missed that

eugeneius commented:
Thank you @Bodacious!

p8 commented:
Thanks @Bodacious!

Regarding this comment:
> > Can't we fix this problem by defining the flash methods as private?

> That would be a much nearer solution, but would break existing code that calls those methods on the controller.

https://github.com/rails/rails/pull/33594#issuecomment-425505321

I'm not sure where anyone would call those methods on the controllers...

Bodacious commented:
@p8 It's bad practice, but it's not difficult to imaging some existing applications calling something like `controller.notice` from helpers or views directly, or `self.notice` within a controller.

Making the flash private would be a breaking change for these cases.

p8 commented:
@Bodacious Ah thanks, I hadn't thought about calling `controller.notice` 👍

2called-chaos commented:
To be fair I wasn't aware of controller.notice/alert but is it therefore fact that these are reserved action names? Prior to Rails 7 you could have actions named this way and work this way

    get '/test' => 'test#alert'

This stops working with Rails 7 but in a non-intuitive way, it just claims to not find that action despite it being listed in /rails/info/routes and you really have to dig to figure out why.

[via reddit](https://www.reddit.com/r/rails/comments/s8tiud/rails_7_are_notice_and_alert_actions_now_reserved/)

ghiculescu commented:
FYI this has had an unexpected side effect - see https://github.com/rails/rails/issues/44867
